### ISSUE TITLE
Remote Routers can't be deployed due to serialization error with JSON.NET

### ISSUE THREAD

**Issue by Aaronontheweb** [akkadotnet/akka.net#545]
During the course of implementing #198 and fixing the associated specs I discovered that JSON.NET can't currently deserialize the `SupervisionStrategy` included on any router config other than `FromConfig` that is sent over the write. I'm in the course of implementing the `DaemonMsgCreateSerializerSpecs` which should help narrow down and resolve this issue.

**Comment by Aaronontheweb** [comment]
I'm currently working on this - can't wrap up my PR for #198 without fixing it.

**Comment by rogeralsing** [comment]
Hmmm.. how could we possibly send the supervisor strategy over the wire?
its code, e.g. a lambda.  right?

**Comment by Horusiath** [comment]
@rogeralsing Does custom supervisor strategy class within shared assembly and default constructor sounds like an option? AFAIK lambda expressions in C# are not fully serializable.

**Comment by rogeralsing** [comment]
Do JVM Akka really do this? or is the supervisor strategy of Props our own invention?
I have a vague feeling that it is.

**Comment by rogeralsing** [comment]
NVM, it's on the RemoteRouterConfig? 
http://doc.akka.io/api/akka/2.3.1/index.html#akka.actor.SupervisorStrategy

So a SupervisionStrategy object needs to be passed over the wire, including the patternmatch for what to do with the different exceptions?

**Comment by Aaronontheweb** [comment]
@rogeralsing correct - we have the `Decider` delegate do this under the hood.

In the short run, just so we get remote router deployment working in _some_ form I would propose having the Deserializer instantiate the default implementation of each type of strategy. Does that sound like more work than serializing / deserializing the `Decider` ?

**Comment by Aaronontheweb** [comment]
Forgot to mention this: I think @Horusiath has the right idea... Looks like what we really need is to provide JSON.NET with a default constructor for deserialization purposes.

**Comment by rogeralsing** [comment]
@Aaronontheweb Sounds good to me.

I have a suggestion form how to replace the Decider delegate:

We add an extension to the `Directive ` enum:

```csharp
public static DirectiveTypePair When<TException>(this Directive self) where TException : Exception{
        return new DirectiveTypePair(self,typeof(TException));
}
```

Then we add an api to take an arglist like this:

```csharp
CreateSupervisorStrategy( Directive defaultDirective, params DirectiveTypePair[] pairs){
```

Then it can be configured like so:
```
 CreateSupervisorStrategy( 
Directive.Stop,  //default
Directive.Escalate.When<NullRefExeption>(), 
Directive.Resule.When<FooException>() );
```

This way, we no longer need a delegate for the supervisor and the list of pairs can easily be serialized...

**Comment by rogeralsing** [comment]
(which means that @HCanber was right all along as usual.. by suggesting this kind of design waaay back)

**Comment by Aaronontheweb** [comment]
Yes he was.

**Comment by Aaronontheweb** [comment]
Although, we can't totally replace the `Decider` delegate - there are instances where the `SupervisionStrategy` gets used to do things in addition to deciding how to handle the exception, such as publishing a child failure to a subscriber and so forth. Custom deciders also get used a lot inside unit tests too.

I'm almost done with the `DaemonMsgCreateSerializerSpec` - there's a unit test in there that's going to fail and will show exactly how much stuff has to get serialized. It's going to be tricky :\

**Comment by Aaronontheweb** [comment]
This is now fixed per #599, although we do need to make this explicit in the documentation somewhere
